### DISCUSSION TITLE
Nested models and default values

### DISCUSSION THREAD

**Discussion by pgiraud** [pydantic/pydantic#2612] [Question] [answered]
With the following code,
```python
from pydantic import BaseModel


class Foo(BaseModel):
    name = 'Jane Doe'
    nick: str


class Spam(BaseModel):
    foo: Foo = Foo(nick="nonick")


Spam.parse_obj({'foo': {'name': "Peter Pan"}})
```

I get the following error:
```
pydantic.error_wrappers.ValidationError: 1 validation error for Spam
foo -> nick
```

While I fully understand the error, what I would like to achieve is to set a value for the `name` property without having to give a value for `nick`. I would like the value for `nick` to be retrieved from the Spam::foo default property.

Do you have any suggestions?

Note: I tried to simplify the use case as much as possible but my real use case it managing nested `Settings` with `_secrets_dir` being dynamic.

**Comment by SvenPVoigt** [comment]
@pgiraud You are getting that error because you are overwriting the default, and the value you are overwriting with does not have a nick attribute (required). Are you not able to set the default in Foo? See below.

```python
class Foo(BaseModel):
    name = 'Jane Doe'
    nick: str = "nonick"
```

Note that `Spam()` constructs a model instance with the default nick value of nonick that you specified.

**Comment by PrettyWood** [comment, marked as answer]
@pgiraud I feel like you want to do something like this
```py
from pydantic import BaseModel, validator


class Foo(BaseModel):
    name = 'Jane Doe'
    nick: str


class Spam(BaseModel):
    foo: Foo = Foo(nick="nonick")

    @validator('foo', pre=True)
    def set_nick_default(cls, v, values, field) -> dict[str, str]:
        assert v == {'name': 'Peter Pan'}
        assert values == {}
        assert field.default == Foo(nick='nonick', name='Jane Doe')
        return {**field.default.dict(), **v}

print(repr(Spam.parse_obj({'foo': {'name': "Peter Pan"}})))
# Spam(foo=Foo(nick='nonick', name='Peter Pan'))
```

**Reply by pgiraud** [reply]
Using `@validator` `pre` was the key. This is exactly what I need.
Thanks a lot for your answer.
